File: frontend/src/lib/urlValidation.ts
export function validateSingleWebsiteUrl(value: string): string | null {
  return validateSingleUrl(value, { requireYoutube: false });
}

export function validateSingleYoutubeUrl(value: string): string | null {
  return validateSingleUrl(value, { requireYoutube: true });
}

function validateSingleUrl(
  value: string,
  options: { requireYoutube: boolean },
): string | null {
  const trimmed = value.trim();
  if (!trimmed) {
    return null;
  }

  const parts = trimmed.split(/\s+/).filter(Boolean);
  if (parts.length !== 1) {
    return null;
  }

  const candidate = parts[0];

  try {
    const parsed = new URL(candidate);
    if (!["http:", "https:"].includes(parsed.protocol) || !parsed.hostname) {
      return null;
    }

    if (options.requireYoutube && !isYoutubeHost(parsed.hostname)) {
      return null;
    }

    return candidate;
  } catch {
    return null;
  }
}

function isYoutubeHost(hostname: string) {
  const normalized = hostname.toLowerCase();
  return (
    normalized === "youtu.be" ||
    normalized === "youtube.com" ||
    normalized === "www.youtube.com" ||
    normalized === "m.youtube.com"
  );
}

File: backend/app/services/url_validation.py
from __future__ import annotations

from urllib.parse import urlparse


def validate_single_website_url(value: str | None) -> str | None:
    return _validate_single_url(value=value, require_youtube=False)


def validate_single_youtube_url(value: str | None) -> str | None:
    return _validate_single_url(value=value, require_youtube=True)


def _validate_single_url(*, value: str | None, require_youtube: bool) -> str | None:
    if value is None:
        return None

    trimmed = value.strip()
    if not trimmed:
        return None

    parts = trimmed.split()
    if len(parts) != 1:
        return None

    candidate = parts[0]
    parsed = urlparse(candidate)

    if parsed.scheme not in {"http", "https"} or not parsed.netloc:
        return None

    if require_youtube and parsed.hostname not in {
        "youtu.be",
        "youtube.com",
        "www.youtube.com",
        "m.youtube.com",
    }:
        return None

    return candidate
